Feat/config

lorenzolewis opened the pull request:
Still a few minor tweaks I need to do, but would like to get a few pairs of eyes on this.

## Items to fix in source:

- [x] "HttpAllowlistScope" description
- [x] Fix link in ShellAllowlistConfig.sidecar description

## Remaining items:

- [x] Link anchor in table
- [x] Required values in properties
- [x] Add value formats
- [x] Maybe revise the example config a bit (include something in the `beforeBuildCommand`, `beforeDevCommand`, etc.)
- [x] Elaborate on "Platform-Specific Configuration" section (would like some help with this)
- [x] Elaborate on 4 main objects listed under "Configuration Structure"
- [x] Check bundle targets
- [x] Remove the `replaceAll` filters at the top after source is fixed
- [x] Format default values a little better
- [x] Add if a value is optional (and a note about this)
- [x] ShellAllowedArgs is nested too deeply
- [x] If a default value is a JS object (DebConfig.files)
- [x] MacConfig, WindowsConfig, SecurityConfig, WindowConfig "any of the following values can be used" troubleshooting
- [x] Is PatternKind, Theme supposed to have content?
- [x] Is UpdaterEndpoint used anywhere?

lorenzolewis committed:
Generate tables

lorenzolewis committed:
Continue improving script

lorenzolewis committed:
Revise template

lorenzolewis committed:
Move bootstrapper info script

lorenzolewis committed:
Fix typo

lorenzolewis committed:
Add required logic

lorenzolewis committed:
Remove unneeded theme components

lorenzolewis committed:
Fix more values

lorenzolewis committed:
Remove logging

lorenzolewis committed:
Remove bootstrapper script per Fabian

lorenzolewis committed:
Merge remote-tracking branch 'origin/dev' into feat/config

lorenzolewis committed:
Fix type for enums and add \n between tables

FabianLars commented on .scripts/generate-config-doc.js:
```diff
@@ -1,149 +1,353 @@
+const schemaString = fs
+  .readFileSync(schemaPath)
+  .toString()
+  .replaceAll('<', '&lt;') // Fixes opening HTML brackets
+  .replaceAll('>', '&gt;') // Fixes closing HTML brackets
```
These ones actually break most of the usages since they are in "backtick blocks". Tricky to to handle both cases here 🤔 
Not counting links, only 2 occurrences without backticks.

lorenzolewis replied:
@lucasfernog This was mainly around the `<file>` items in the schema because it was being recognised as HTML tags. I think if those are removed in the schema then this can be removed

FabianLars replied:
yeah that's the 2 occurrences i was talking about :P

FabianLars commented:
I feel like the headers of the allowlist objects are too small, especially being next to the big tables. They're too easy to miss rn.

FabianLars commented:
The Scope objects are really weird too, but i don't really know how to integrate them better without flattening them into the respective allowlist config table (actually i'd like that, but _no_ idea how to do it.

Edit: Or maybe they should get separate pages like in the js docs 🤔

lucasfernog committed:
filter 'null' from non-required types

lorenzolewis commented:
> I feel like the headers of the allowlist objects are too small, especially being next to the big tables. They're too easy to miss rn.

I could set it to where the minimum heading size is an h5 instead of h6 to help with that

> Edit: Or maybe they should get separate pages like in the js docs 🤔

I've been looking at using an automatic generator based on the schema (https://github.com/adobe/jsonschema2md) and it could do this, but we need to do some TLC on https://github.com/tauri-apps/tauri/issues/4139 first

lorenzolewis committed:
Don't allow super small headings

lorenzolewis commented:
Just did a push for the headings, lmk what you think

FabianLars commented:
I'm talking about these:
![grafik](https://user-images.githubusercontent.com/30730186/168589631-5ba9c008-511b-43a4-93a6-bfebd5762d05.png)

And they are already h4, but imo they don't stand out enough to make clear that they're the headers of the tables

lucasfernog committed:
fix: do not replace < and >

lorenzolewis commented:
> I'm talking about these: ![grafik](https://user-images.githubusercontent.com/30730186/168589631-5ba9c008-511b-43a4-93a6-bfebd5762d05.png)
> 
> And they are already h4, but imo they don't stand out enough to make clear that they're the headers of the tables

@FabianLars how would you feel about leaving this as-is if we look at something like https://github.com/adobe/jsonschema2md in the future that can automatically split things to separate pages? Either that or we have h3 be the minimum heading size, but then that promotes every object up to the overview bar on the right side.

lorenzolewis commented:
For the scope objects, I don't quote know how to render "anyOf" type of properties (those are the types that generate the below type of table). But at least now it's rendered, in the current live config these aren't even rendered 😛 
<img width="917" alt="Screenshot 2022-05-16 at 13 56 33" src="https://user-images.githubusercontent.com/15347255/168597474-644ca5fb-23ba-49b3-b6d5-954623ca80e9.png">

lucasfernog committed:
fix: parse local urls

lucasfernog committed:
fix newlines, examples in description

lucasfernog committed:
update link

FabianLars commented:
> how would you feel about leaving this as-is if we look at something like https://github.com/adobe/jsonschema2md in the future that can automatically split things to separate pages? Either that or we have h3 be the minimum heading size, but then that promotes every object up to the overview bar on the right side.

Can we give them `font-size: 1.2rem;` or something for now?

> For the scope objects, I don't quote know how to render "anyOf" ...

Yeah the tables are alright-ish, but the text above is a little weird? idk how to explain it, but maybe it's connected to the small titles too.

Another thing i can't really explain: 
![grafik](https://user-images.githubusercontent.com/30730186/168604777-f687574a-ee3e-4c8a-bfdf-d5c6672af33d.png)

Like the 2 "Any of the following..." don't make sense there at all and i can't really see where they even come from; probably a scheme issue tho?

lucasfernog committed:
feat: add anchor to properties in table

lorenzolewis committed:
Adapt anyOf

lorenzolewis committed:
Merge remote-tracking branch 'origin/dev' into feat/config

lorenzolewis committed:
Clean up anyOf code, change oneOf to not render table

lorenzolewis commented:
Changed anchor to be on the same line of properties in the table and then I changed the format of anyOf and oneOf so it's a bullet list instead of a table now (seems to make more sense).

lorenzolewis committed:
Simplify format

lorenzolewis commented:
It was bothering me so I overhauled it. Technically having the hierarchy like it was isn't a correct schema because a "definition" can relate to multiple objects (and was causing a lot of issues). So although this is "structured" as cleanly, it's _more correct_.

Just a few minor things I want to work on that I put in the description, but let me know what you think

lorenzolewis committed:
Workaround for broken link

lorenzolewis committed:
Escape any angle brackets

lucasfernog committed:
readd description constructor, fix `-` formatting

lucasfernog committed:
remove template, format json

lorenzolewis committed:
Prettier formatting, render child objects as an h3

lorenzolewis committed:
Install CI deps

lorenzolewis committed:
Switch to using `JSON` instead of `JSON5`

lorenzolewis committed:
Temp fix for link capitalisation

lorenzolewis committed:
Add type format

lorenzolewis committed:
Add "required" logic

lorenzolewis committed:
Test upstream preserve order

lorenzolewis committed:
Add back line break fix

lorenzolewis committed:
Add comments to script [skip ci]

lorenzolewis committed:
Update description order [skip ci]

lorenzolewis committed:
Fix anchor padding [skip-ci]

lorenzolewis commented:
Ready to merge as soon as https://github.com/tauri-apps/tauri/pull/4153 is merged

lucasfernog committed:
fix: remove useless replace

lucasfernog reviewed (approved):
This is just amazing!!!

lorenzolewis commented:
@FabianLars you have a requested change open on this still

lorenzolewis commented:
> This is just amazing!!!

Thanks! Took me 4 iterations and 2 re-writes, but I got there in the end! :D

lorenzolewis committed:
h3 spacing so Fabian can have a good day :)

FabianLars reviewed (approved):
LGTM (for now :P )

lorenzolewis commented:
Quick, merge before he changes his mind!
